README.md
# MileageStats

Angular 2 implementation of Mileage Stats application that is part of Project Silk of Microsoft. Application is coded from the grounds up without using code or patterns from this project; only images are imported.

## Usage

You need to have node.js and mongoDB installed on your system. Clone the repository and go into `/server` directory. Type `node init-script.js`. This will initialize the database. Now from the root directory start the app by typing `node server` and go to `localhost:3000` from the browser. You can use user name `sample` to log in with sample data.

## Source code

`/src/app` directory contains the client source code, while rudimentory back-end code I wrote you can find in `/server` directory. Node server itself is in `./server.js` file.
Source code documentation generated with YUIdoc is available in `/docs` directory.
Production bundle used by the actual application is in `/public` directory.
Long running Git branch for further development is `dev`.

## Angular-CLI

This project was generated with [angular-cli](https://github.com/angular/angular-cli) version 1.1.3.

### Development server
Checkout the `dev` branch and run `ng serve` for a dev server. In another console start development node backend with `node server-dev`. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.

### Code scaffolding

Run `ng generate component component-name` to generate a new component. You can also use `ng generate directive/pipe/service/class`.

### Build

Run `ng build` to build the project. The build artifacts will be stored in the `dist/` directory. Use the `-prod` flag for a production build.

### Running unit tests

Run `ng test` to execute the unit tests via [Karma](https://karma-runner.github.io).

### Running end-to-end tests

Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).
Before running the tests make sure you are serving the app via `ng serve`.

### Deploying to Github Pages

Run `ng github-pages:deploy` to deploy to Github Pages.

### Further help

To get more help on the `angular-cli` use `ng --help` or go check out the [Angular-CLI README](https://github.com/angular/angular-cli/blob/master/README.md).


CHANGELOG.txt
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]


## [1.1.1] - 2017-11-10
### Changed
- Focused element is made more visible by panting red border around it
- Application is now compiled ahead of time and starts much faster
### Fixed
- Sign-out now properly receives focus


## [1.1] - 2017-09-23
### Added
- Drop down menu for narrow resolutions
### Changed
- Documentation now includes modules
### Fixed
- onBlur text trimming removed due to unwanted triggering of the  validators


## [1.0] - 2017-07-22
### Added
- Animation on routing
### Changed
- e2e tests start from landing/signin page, entering 'sample' in username field
### Fixed
- Esc quit key bug.
- Sign up component observables bug.
- A bug when cancelling car editing fixed.
- Complete your registration form now dissapears after hitting the save button.


##[0.9.0] - 2017-07-22
### Added
- Summary pane.


##[0.8.0] - 2017-07-22
### Added
- Authentication.


## [0.7.0] - 2017-07-22
### Added
- Form inputs stylized
- Remindres stylized
- Fill ups sylized
- Basic car data stylized
- Accordion componentn stylized
- Car cards sylized
- Status bar sylized


## [0.6.1] - 2017-07-21
### Added
- Individual fill ups and reminders can now be selected using keyboard.
### Fixed
- Efficiency when redrawing accordion widget on resolution change.
- Unit and black-box tests corrected to work properly.


## [0.6.0] - 2017-07-20
### Added
- Adding and changing car picture, with cropping ability
- Node server and database client added
- Server written and tested
### Fixed
- Appropriate car card is selected after switching to details view (no car card was being marked as selected before)


## [0.5.0] - 2017-07-18
### Added
- Responsive layout with four break-points
- Responsive sizing of all the elements
- Overdue reminders in basic car data component
- Highcharts charts component


## [0.4.0] - 2017-07-09
### Added
- Sorting fill ups and reminders list.
- *cars-list* and *car-details-view*
- Deleting a car
- Editing a car
- Editing and deleting a fill up and a reminder
- Trimming input fields and canceling input via cancel button or esc key.
- Adding a reminder.
- Showing reminders.
- Code commented and formatted.
- Redux wired in.
- Utlilities service-error handling
- Notification service
### Changed
- CarCards component now uses resolvers to fetch data.
- Accordion widget further developed.


## [0.3.0] - 2017-07-07
### Added
- *fill-up-detail* - display of single fill up details implemented
- *fill-ups-list*, *fill-ups-list-entry* - fill ups list basically implemented.
- *basic-car-data-, fill-ups, reminders* - these views wired into accordion component.
- *accordion.component.ts* - basic functionality of accordion widget.
### Fixed
- DataService is now returning correct data on add, update and delete operations.


## [0.2.0] - 2017-07-05
### Added
- Form validation for AddCar implemented.
- *cars-data-resolver* implemented.
- Routing module introduced. Rest of app modularized into Dasboard and AddCar modules.
- *add-car.component.ts* AddCar component basics implemented.
- routing.
- *add-car-button.component.ts* AddCarButton component.


## [0.1.0] - 2017-07-03
### Added
- *car-card.component.ts* CarCardsComponent which shows all car cards basically finished
- *car-card.component.ts* CarCardComponent which shows a single car card basically finished
- DataService class completed
- *data.service.ts* DataService class which handles HTTP requests towards the server; getCars(), getFillUps() and getReminders() methods implemented together with a belonging test
- *in-memory-data-service.ts* InMemoryDataService class which will stub the back end using InMemoryDbService.

server/init-script.js
/* This module creates initial database for sample user */

"use strict";

var mongoose = require( 'mongoose' );
var fs=require('fs');

var mongoDBConfig = require('./MongoDBConfig');

// Create the database connection 
mongoose.connect(mongoDBConfig.dbURI);

// CONNECTION EVENTS
// When successfully connected
mongoose.connection.on('connected', function() {
	console.log('Mongoose default connection open to ' + mongoDBConfig.dbURI);
});

var MongoUser = require('./MongoUser');
var MongoCar = require('./MongoCar');
var MongoFillUp = require('./MongoFillUp');
var MongoReminder = require('./MongoReminder');
var MongoCarImage = require('./MongoCarImage');
/*
 */

var carImage = new MongoCarImage({
	data: fs.readFileSync('../src/assets/images/vehicle.png'),
	contentType: 'image/png',
	name: "placeholder",
	tmpPath: __dirname + "../src/assets/images/vehicle.png"
});
carImage.save(function(err){
	if (err) console.log(err);
});

var fillUps = [
	[	new MongoFillUp({quantity: 11, pricePerGalon: 2.81, date:'2017-04-28', station: 'Fabrikam', odometer: 15455}),
		new MongoFillUp({quantity: 7, pricePerGalon: 3, date:'2017-06-12', station: 'Joe', odometer: 15843}),
		new MongoFillUp({quantity: 20, pricePerGalon: 3, date:'2017-07-12', station: 'Miriam', odometer: 16000}),
	],
	[ new MongoFillUp({quantity: 11, pricePerGalon: 2.81, date:'2017-02-28', station: 'Fabrikam', odometer: 15455}),
		new MongoFillUp({quantity: 15, pricePerGalon: 2, date:'2017-03-28', station: 'Fabrikam', odometer: 15923}),
		new MongoFillUp({quantity: 8, pricePerGalon: 2.81, date:'2017-04-28', station: 'Fabrikam', odometer: 16213}),
		new MongoFillUp({quantity: 23, pricePerGalon: 1.7, date:'2017-06-28', station: 'Fabrikam', odometer: 16455})
	],
	[	new MongoFillUp({quantity: 11, pricePerGalon: 2.81, date:'2017-01-28', station: 'Fabrikam', odometer: 15455}),
		new MongoFillUp({quantity: 7, pricePerGalon: 3, date:'2017-02-12', station: 'Joe', odometer: 15843}),
		new MongoFillUp({quantity: 10, pricePerGalon: 2.81, date:'2017-04-12', station: 'Fabrikam', odometer: 16223}),
		new MongoFillUp({quantity: 4, pricePerGalon: 2.61, date:'2017-05-23', station: 'Cheepo', odometer: 16589}),
		new MongoFillUp({quantity: 24, pricePerGalon: 2, date:'2017-06-23', station: 'Cheepo', odometer: 16678})
	]
];
var reminders = [
	[ new MongoReminder({text: 'Check Fluid', date:'2030-05-07'})
	],
	[ new MongoReminder({text: 'Check Fluid', date:'2016-05-07'}),
	  new MongoReminder({text: 'Vacuum Car', date:'2016-06-03'}),
    new MongoReminder({text: 'Change Tires', date:'2016-09-23'}),
    new MongoReminder({text: 'Vacuum Car', date:'2016-10-01'}),
    new MongoReminder({text: 'Check Fluid', date:'2030-11-05'})
	],
	[ new MongoReminder({text: 'Check Fluid', date:'2016-05-07'}),
		new MongoReminder({text: 'Change Tires', date:'2016-09-23'}),
		new MongoReminder({text: 'Check Fluid', date:'2016-11-05'})
	]
];
var cars = [
	new MongoCar({name: 'Hot Rod', model: 'BMW 330xi', year:'2003', avatar: carImage}),
	new MongoCar({name: "Soccer Mom's Ride", model: 'Honda Accord LX', year: '1997', avatar: carImage}),
	new MongoCar({name: 'Mud Lover', model: 'Jeep Wrangler', year: '2011', avatar: carImage}) 
];

var user = new MongoUser({username: 'sample'});

/*
*/
  for (var i = 0; i < cars.length; ++i) {
    for (var j = 0; j < fillUps[i].length; ++j) {
      cars[i].fillUps.push(fillUps[i][j]);
      fillUps[i][j].save();
    }
    for (j = 0; j < reminders[i].length; ++j) {
      cars[i].reminders.push(reminders[i][j]);
      reminders[i][j].save(function (err) {
				if (err) console.log(err);
			});
    }
    user.cars.push(cars[i]);
    cars[i].save();
  }
  
  user.save(function (err) {
		if (err) console.log(err);
	});

	user = new MongoUser({username: 'dalibor'});
	user.save(function (err) {
		if (err) console.log(err);
	});

// If the connection throws an error
mongoose.connection.on('error', function(err) {
	console.log('Mongoose default connection error: ' + err);
});

// When the connection is disconnected
mongoose.connection.on('disconnected', function () {
	console.log('Mongoose default connection disconnected');
} );

// When Node server exits
process.on('SIGINT', function() {
	mongoose.connection.close(function () {
		console.log('Mongoose disconnected through app termination');
		process.exit(0);
	});
});

mongoose.connection.close();